### DISCUSSION TITLE
How to remove the ‘src’ folder in ‘node_modules/@theia/*’ of theia-blueprint packaged version？

### DISCUSSION THREAD

**Discussion by inlann** [eclipse-theia/theia#10654] [General]
Hello:

I posted a [question](https://community.theia-ide.org/t/how-to-remove-the-src-folder-in-node-modules-theia-of-theia-blueprint-packaged-version/2185) in Theia community but I'm not sure if I need to post it here too. So, I copied it below. Sorry to bother.

We found that the unpacked version of Theia-blueprint still kept Theia’s source code in its `resources/app/node_module/@theia/*` folder, which would leak the source code and increase the size of application:

```bash
tptuser@linux-3:/work/ide/theia-blueprint/applications/electron/dist/linux-unpacked/resources/app/node_modules/@theia$ ls */src/*
application-package/src/api.ts                          external-terminal/src/package.spec.ts      node-pty/src/types.ts                 plugin-ext/src/plugin-ext-backend-electron-module.ts
application-package/src/application-package.spec.ts     getting-started/src/package.spec.ts        node-pty/src/unixTerminal.test.ts     plugin-ext/src/plugin-ext-backend-module.ts
application-package/src/application-package.ts          keymaps/src/package.spec.ts                node-pty/src/unixTerminal.ts          plugin-ext/src/plugin-ext-frontend-electron-module.ts
application-package/src/application-props.ts            mini-browser/src/package.spec.ts           node-pty/src/utils.ts                 plugin-ext/src/plugin-ext-frontend-module.ts
application-package/src/environment.ts                  monaco/src/package.spec.ts                 node-pty/src/windowsPtyAgent.test.ts  plugin/src/package.spec.ts
application-package/src/extension-package-collector.ts  navigator/src/package.spec.ts              node-pty/src/windowsPtyAgent.ts       property-view/src/package.spec.ts
application-package/src/extension-package.ts            node-pty/src/conpty_console_list_agent.ts  node-pty/src/windowsTerminal.test.ts  terminal/src/package.spec.ts
application-package/src/index.ts                        node-pty/src/eventEmitter2.test.ts         node-pty/src/windowsTerminal.ts       timeline/src/package.spec.ts
application-package/src/json-file.ts                    node-pty/src/eventEmitter2.ts              outline-view/src/package.spec.ts      typehierarchy/src/package.spec.ts
application-package/src/npm-registry.ts                 node-pty/src/index.ts                      ovsx-client/src/index.ts              userstorage/src/package.spec.ts
callhierarchy/src/package.spec.ts                       node-pty/src/interfaces.ts                 ovsx-client/src/ovsx-client.spec.ts   vsx-registry/src/package.spec.ts
console/src/package.spec.ts                             node-pty/src/terminal.test.ts              ovsx-client/src/ovsx-client.ts
editor-preview/src/package.spec.ts                      node-pty/src/terminal.ts                   ovsx-client/src/ovsx-types.ts
editor/src/package.spec.ts                              node-pty/src/testUtils.test.ts             plugin-dev/src/package.spec.ts

bulk-edit/src/browser:
bulk-edit-commands.ts  bulk-edit-contribution.ts  bulk-edit-frontend-module.ts  bulk-edit-tree  bulk-edit-tree-label-provider.ts  style

callhierarchy/src/browser:
callhierarchy-contribution.ts  callhierarchy-frontend-module.ts  callhierarchy-service.ts  callhierarchy-tree  callhierarchy.ts  current-editor-access.ts  index.ts  style  utils.ts

callhierarchy/src/common:
glob.ts  language-selector.ts  paths.ts
```

**Is there any way to let electron-builder to remove the src folder?**

**Another question**: If we let electron-builder to remove the src folder for each theia module in node_module/@theia, the style file imports would have trouble because the src folder is no longer in the packaged version:

```typescript
# packages/core/src/browser/frontend-application-module.ts

import '../../src/browser/style/index.css';
require('../../src/browser/style/materialcolors.css').use();
import 'font-awesome/css/font-awesome.min.css';
import 'file-icons-js/css/style.css';
import '@vscode/codicons/dist/codicon.css';

// ...
```

**Is there any way to copy style files from src to lib so that we could import it using: import './style/index.css' ?**

Really appreciate :smiley:

**Comment by msujew** [comment]
Hi @inlann,

Sorry for not getting back to you on the community forum, posts on there sometimes go unnoticed (by me at least). Since the `src` paths are hardcoded in Theia, I don't think it would be feasible to delete the whole `src` folder for every Theia package.

Instead of simply ignoring/removing the `node_modules/@theia` folder, would it be an option to perform a prepackage build script? A simple JavaScript file which uses something like `glob` to identify all `.ts` source files in the `node_modules/@theia` folders and delete them all in one go? I think that would be a feasible way to deal with this issue.

On another note: I think these css import statements will be already resolved once the `webpack` build has finished, so you should be save to delete the whole `src` folder afterwards.

**Reply by inlann** [reply]
Hello @msujew,

Really thanks for response 😄 

>Instead of simply ignoring/removing the node_modules/@theia folder, would it be an option to perform a prepackage build script? A simple JavaScript file which uses something like glob to identify all .ts source files in the node_modules/@theia folders and delete them all in one go? I think that would be a feasible way to deal with this issue.

You're right. We currently use a shell script to delete the .ts files in the `node_modules/@theia/*/src` directory as a result of each `yarn package` execution. Although it's only a solution for the linux-unpacked version because we can't modify the files of the AppImage or .deb package, but the linux-unpacked version is enough for us at the moment.

About the import of style files, I will test whether removing the entire src directory will cause styles invalid in the packaged version (`webpack` build finished). 😄

**Comment by kittaakos** [comment]
> to remove the src folder?

The [arduino-ide](https://github.com/arduino/arduino-ide/) removes the Theia `.ts`, `spec.js`, etc files like [this](https://github.com/arduino/arduino-ide/blob/1d88263c853cd9a431ff580227a9e5bb8edd64e5/electron/build/template-package.json#L57-L70).

> the style file imports would have trouble because the src folder is no longer in the packaged version:

In theory, anything under `node_modules/@theia/package-name/src/browser` can be removed from the final electron app. All the browser code, style, assets, etc., should be picked by `webpack` and will be somewhere in the output folder, where your `bundle.js` is.

Also, the `arduino-ide` [collects unused dependencies](https://github.com/arduino/arduino-ide/blob/1d88263c853cd9a431ff580227a9e5bb8edd64e5/electron/packager/index.js#L113-L116) and filters them too. Maybe it's not possible with a more recent Theia version. I do not know. Back then, I did not hear of any complaints/bug reports due to this kind of "optimization". I hope this helps, @inlann. 🤞

**Reply by inlann** [reply]
It's really helpful!! 😍

The [files field](https://github.com/arduino/arduino-ide/blob/1d88263c853cd9a431ff580227a9e5bb8edd64e5/electron/build/template-package.json#L57-L70) of arduino-ide :

```
    "files": [
      "src-gen",
      "lib",
      "scripts",
      "!node_modules/**/*.{ts,map}",
      "!node_modules/**/*.spec.js",
      "!node_modules/@theia/**/test/*",
      "!node_modules/@theia/**/src/*.ts",
      "!node_modules/@theia/**/lib/*browser/*",
      "node_modules/@theia/core/lib/browser/*",
      "!node_modules/@typefox/monaco-editor-core/*",
      "!node_modules/oniguruma/*",
      "!node_modules/onigasm/*"
    ],
```

The line `"!node_modules/@theia/**/src/*.ts",` maybe need to be modified to `"!node_modules/@theia/**/src/**/*.ts",` because it only filters the `*.ts` files under the `src` folder but there are some nested folders under the `src` like `node_modules/@theia/bulk-edit/src/browser/bulk-edit-tree`.

And maybe need a new line `"!node_modules/@theia/**/src/**/*.tsx",` to filter `*.tsx` files too. 😄
